File: apps/api/src/modules/auth/service/auth-service.ts
import {
  AppError,
  type AuthSessionResponse,
  type LoginRequest,
  type RegisterRequest,
} from "@repo/contracts";
import * as argon2 from "argon2";
import {
  createRefreshTokenValue,
  hashToken,
  signAccessToken,
  verifyAccessToken,
} from "../../../lib/tokens.js";
import { toAuthUser } from "../types/auth-types.js";

export type AuthUserRow = {
  id: string;
  email: string;
  passwordHash: string;
  role: "user" | "admin";
  createdAt: Date;
};

export type AuthUserStore = {
  findByEmail: (email: string) => Promise<AuthUserRow | null>;
  findById: (id: string) => Promise<AuthUserRow | null>;
  create: (input: { email: string; passwordHash: string }) => Promise<AuthUserRow>;
};

export type AuthRefreshStore = {
  create: (input: { userId: string; tokenHash: string; expiresAt: Date }) => Promise<unknown>;
  findByHash: (tokenHash: string) => Promise<{
    tokenHash: string;
    expiresAt: Date;
    user: AuthUserRow;
  } | null>;
  /** Atomically take ownership of a refresh session (delete + return), or null if missing/lost race. */
  consumeByHash: (tokenHash: string) => Promise<{
    tokenHash: string;
    expiresAt: Date;
    user: AuthUserRow;
  } | null>;
  deleteByHash: (tokenHash: string) => Promise<unknown>;
};

export type AuthIdempotencyStore = {
  findByKey: (key: string) => Promise<{ responseBody: unknown } | null>;
  create: (input: {
    key: string;
    method: string;
    path: string;
    statusCode: number;
    responseBody: unknown;
  }) => Promise<unknown>;
};

export type AuthStores = {
  users: AuthUserStore;
  refresh: AuthRefreshStore;
  idempotency: AuthIdempotencyStore;
};

export type AuthUnitOfWork = {
  run: <T>(fn: (stores: AuthStores) => Promise<T>) => Promise<T>;
};

export type AuthServiceDeps = {
  users: AuthUserStore;
  refresh: AuthRefreshStore;
  idempotency: AuthIdempotencyStore;
  /** When provided, register/refresh multi-writes run inside this unit of work. */
  uow?: AuthUnitOfWork;
  /** Map persistence unique violations to domain fields (email | key). */
  isUniqueViolation?: (error: unknown, field: "email" | "key") => boolean;
  jwtSecret: string;
  now?: () => Date;
};

export type AuthTokens = {
  accessToken: string;
  refreshToken: string;
};

export type AuthResult = {
  session: AuthSessionResponse;
  tokens: AuthTokens;
};

type TokenUser = {
  id: string;
  email: string;
  role: "user" | "admin";
  createdAt: Date;
};

const defaultIsUniqueViolation: NonNullable<AuthServiceDeps["isUniqueViolation"]> = () => false;

export function createAuthService(deps: AuthServiceDeps) {
  const now = deps.now ?? (() => new Date());
  const isUniqueViolation = deps.isUniqueViolation ?? defaultIsUniqueViolation;
  const rootStores: AuthStores = {
    users: deps.users,
    refresh: deps.refresh,
    idempotency: deps.idempotency,
  };
  const uow: AuthUnitOfWork = deps.uow ?? {
    run: async (fn) => fn(rootStores),
  };

  async function issueTokens(refreshStore: AuthRefreshStore, user: TokenUser): Promise<AuthResult> {
    const accessToken = await signAccessToken(
      { sub: user.id, email: user.email, role: user.role },
      deps.jwtSecret,
    );
    const refreshToken = createRefreshTokenValue();
    const expiresAt = new Date(now().getTime() + 1000 * 60 * 60 * 24 * 30);
    await refreshStore.create({
      userId: user.id,
      tokenHash: hashToken(refreshToken),
      expiresAt,
    });
    return {
      session: {
        user: {
          id: user.id,
          email: user.email,
          role: user.role,
          createdAt: user.createdAt.toISOString(),
        },
      },
      tokens: { accessToken, refreshToken },
    };
  }

  async function replayIdempotentRegister(
    storedBody: unknown,
    input: RegisterRequest,
  ): Promise<AuthResult> {
    const stored = storedBody as AuthSessionResponse;
    const user = await deps.users.findById(stored.user.id);
    if (!user) {
      throw new AppError({
        status: 409,
        errorCode: "IDEMPOTENCY_CONFLICT",
        errorMessage: "Idempotent replay failed",
      });
    }
    const emailMatches = user.email.toLowerCase() === input.email.toLowerCase();
    const passwordMatches = await argon2.verify(user.passwordHash, input.password);
    if (!emailMatches || !passwordMatches) {
      throw new AppError({
        status: 409,
        errorCode: "IDEMPOTENCY_CONFLICT",
        errorMessage: "Idempotency key reused with different credentials",
      });
    }
    return issueTokens(deps.refresh, user);
  }

  return {
    async register(input: RegisterRequest, idempotencyKey?: string): Promise<AuthResult> {
      if (idempotencyKey) {
        const existing = await deps.idempotency.findByKey(idempotencyKey);
        if (existing) {
          return replayIdempotentRegister(existing.responseBody, input);
        }
      }

      const existingUser = await deps.users.findByEmail(input.email);
      if (existingUser) {
        throw new AppError({
          status: 409,
          errorCode: "AUTH_EMAIL_TAKEN",
          errorMessage: "Email already registered",
        });
      }

      const passwordHash = await argon2.hash(input.password);

      try {
        return await uow.run(async (stores) => {
          const user = await stores.users.create({ email: input.email, passwordHash });
          const result = await issueTokens(stores.refresh, user);

          if (idempotencyKey) {
            await stores.idempotency.create({
              key: idempotencyKey,
              method: "POST",
              path: "/api/v1/auth/register",
              statusCode: 201,
              responseBody: result.session,
            });
          }

          return result;
        });
      } catch (error) {
        if (error instanceof AppError) {
          throw error;
        }
        if (idempotencyKey && isUniqueViolation(error, "key")) {
          const existing = await deps.idempotency.findByKey(idempotencyKey);
          if (existing) {
            return replayIdempotentRegister(existing.responseBody, input);
          }
          throw new AppError({
            status: 409,
            errorCode: "IDEMPOTENCY_CONFLICT",
            errorMessage: "Idempotent replay failed",
          });
        }
        if (isUniqueViolation(error, "email")) {
          // Concurrent same-key register may surface as email unique after the
          // winner commits; prefer credential-verified idempotent replay.
          if (idempotencyKey) {
            const existing = await deps.idempotency.findByKey(idempotencyKey);
            if (existing) {
              return replayIdempotentRegister(existing.responseBody, input);
            }
          }
          throw new AppError({
            status: 409,
            errorCode: "AUTH_EMAIL_TAKEN",
            errorMessage: "Email already registered",
          });
        }
        throw error;
      }
    },

    async login(input: LoginRequest): Promise<AuthResult> {
      const user = await deps.users.findByEmail(input.email);
      if (!user) {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_INVALID_CREDENTIALS",
          errorMessage: "Invalid credentials",
        });
      }

      const valid = await argon2.verify(user.passwordHash, input.password);
      if (!valid) {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_INVALID_CREDENTIALS",
          errorMessage: "Invalid credentials",
        });
      }

      return issueTokens(deps.refresh, user);
    },

    async me(accessToken: string | undefined): Promise<AuthSessionResponse> {
      if (!accessToken) {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_UNAUTHORIZED",
          errorMessage: "Missing access token",
        });
      }

      let payload: Awaited<ReturnType<typeof verifyAccessToken>>;
      try {
        payload = await verifyAccessToken(accessToken, deps.jwtSecret);
      } catch {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_UNAUTHORIZED",
          errorMessage: "Invalid access token",
        });
      }

      const user = await deps.users.findById(payload.sub);
      if (!user) {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_UNAUTHORIZED",
          errorMessage: "User not found",
        });
      }

      return { user: toAuthUser(user) };
    },

    async refresh(refreshToken: string | undefined): Promise<AuthResult> {
      if (!refreshToken) {
        throw new AppError({
          status: 401,
          errorCode: "AUTH_UNAUTHORIZED",
          errorMessage: "Missing refresh token",
        });
      }

      return uow.run(async (stores) => {
        const session = await stores.refresh.consumeByHash(hashToken(refreshToken));
        if (!session || session.expiresAt.getTime() < now().getTime()) {
          throw new AppError({
            status: 401,
            errorCode: "AUTH_UNAUTHORIZED",
            errorMessage: "Invalid refresh token",
          });
        }

        return issueTokens(stores.refresh, session.user);
      });
    },

    async logout(refreshToken: string | undefined): Promise<void> {
      if (refreshToken) {
        await deps.refresh.deleteByHash(hashToken(refreshToken));
      }
    },
  };
}

export type AuthService = ReturnType<typeof createAuthService>;

File: templates/api-python/app/modules/auth/service/auth_service.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any, Protocol

from pwdlib import PasswordHash

from app.lib.contracts_catalog import ERROR_CODES
from app.lib.contracts_models import AuthSessionResponse, LoginRequest, RegisterRequest
from app.lib.errors import AppError
from app.lib.tokens import (
    create_refresh_token_value,
    hash_token,
    sign_access_token,
    verify_access_token,
)

password_hash = PasswordHash.recommended()


def _iso(value: datetime) -> str:
    if value.tzinfo:
        return value.replace(tzinfo=UTC).isoformat().replace("+00:00", "Z")
    return value.isoformat() + "Z"


@dataclass
class StoredUser:
    id: str
    email: str
    password_hash: str
    role: str
    created_at: datetime


class AuthUserStore(Protocol):
    def find_by_email(self, email: str) -> StoredUser | None: ...
    def find_by_id(self, user_id: str) -> StoredUser | None: ...
    def create(self, *, email: str, password_hash: str) -> StoredUser: ...


class AuthRefreshStore(Protocol):
    def create(self, *, user_id: str, token_hash: str, expires_at: datetime) -> None: ...
    def find_by_hash(self, token_hash: str) -> tuple[str, datetime, StoredUser] | None: ...
    def delete_by_hash(self, token_hash: str) -> None: ...


class AuthIdempotencyStore(Protocol):
    def find_by_key(self, key: str) -> dict[str, Any] | None: ...
    def create(
        self,
        *,
        key: str,
        method: str,
        path: str,
        status_code: int,
        response_body: dict[str, Any],
    ) -> None: ...


@dataclass
class AuthTokens:
    access_token: str
    refresh_token: str


@dataclass
class AuthResult:
    session: AuthSessionResponse
    tokens: AuthTokens


@dataclass
class AuthService:
    users: AuthUserStore
    refresh_store: AuthRefreshStore
    idempotency_store: AuthIdempotencyStore
    jwt_secret: str
    now: Callable[[], datetime]

    def issue_tokens(self, user: StoredUser) -> AuthResult:
        access_token = sign_access_token(
            sub=user.id,
            email=user.email,
            role=user.role,  # type: ignore[arg-type]
            secret=self.jwt_secret,
            now=self.now(),
        )
        refresh_token = create_refresh_token_value()
        expires_at = self.now() + timedelta(days=30)
        self.refresh_store.create(
            user_id=user.id,
            token_hash=hash_token(refresh_token),
            expires_at=expires_at,
        )
        session = AuthSessionResponse.model_validate(
            {
                "user": {
                    "id": user.id,
                    "email": user.email,
                    "role": user.role,
                    "createdAt": _iso(user.created_at),
                }
            }
        )
        return AuthResult(
            session=session,
            tokens=AuthTokens(access_token=access_token, refresh_token=refresh_token),
        )

    def register(
        self,
        input_data: RegisterRequest,
        idempotency_key: str | None = None,
    ) -> AuthResult:
        if idempotency_key:
            existing = self.idempotency_store.find_by_key(idempotency_key)
            if existing is not None:
                stored = AuthSessionResponse.model_validate(existing)
                user = self.users.find_by_id(stored.user.id)
                if user is None:
                    raise AppError(
                        status=409,
                        error_code=ERROR_CODES["IDEMPOTENCY_CONFLICT"],
                        error_message="Idempotent replay failed",
                    )
                return self.issue_tokens(user)

        existing_user = self.users.find_by_email(str(input_data.email))
        if existing_user is not None:
            raise AppError(
                status=409,
                error_code=ERROR_CODES["AUTH_EMAIL_TAKEN"],
                error_message="Email already registered",
            )

        hashed = password_hash.hash(input_data.password)
        user = self.users.create(email=str(input_data.email), password_hash=hashed)
        result = self.issue_tokens(user)

        if idempotency_key:
            self.idempotency_store.create(
                key=idempotency_key,
                method="POST",
                path="/api/v1/auth/register",
                status_code=201,
                response_body=result.session.model_dump(by_alias=True),
            )

        return result

    def login(self, input_data: LoginRequest) -> AuthResult:
        user = self.users.find_by_email(str(input_data.email))
        if user is None:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_INVALID_CREDENTIALS"],
                error_message="Invalid credentials",
            )

        if not password_hash.verify(input_data.password, user.password_hash):
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_INVALID_CREDENTIALS"],
                error_message="Invalid credentials",
            )

        return self.issue_tokens(user)

    def me(self, access_token: str | None) -> AuthSessionResponse:
        if not access_token:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="Missing access token",
            )

        try:
            payload = verify_access_token(access_token, self.jwt_secret)
        except Exception as exc:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="Invalid access token",
            ) from exc

        user = self.users.find_by_id(payload["sub"])
        if user is None:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="User not found",
            )

        return AuthSessionResponse.model_validate(
            {
                "user": {
                    "id": user.id,
                    "email": user.email,
                    "role": user.role,
                    "createdAt": _iso(user.created_at),
                }
            }
        )

    def refresh(self, refresh_token: str | None) -> AuthResult:
        if not refresh_token:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="Missing refresh token",
            )

        found = self.refresh_store.find_by_hash(hash_token(refresh_token))
        if found is None:
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="Invalid refresh token",
            )

        token_hash, expires_at, user = found
        if expires_at.replace(tzinfo=UTC) < self.now().replace(tzinfo=UTC):
            raise AppError(
                status=401,
                error_code=ERROR_CODES["AUTH_UNAUTHORIZED"],
                error_message="Invalid refresh token",
            )

        self.refresh_store.delete_by_hash(token_hash)
        return self.issue_tokens(user)

    def logout(self, refresh_token: str | None) -> None:
        if refresh_token:
            self.refresh_store.delete_by_hash(hash_token(refresh_token))


def create_auth_service(
    *,
    users: AuthUserStore,
    refresh: AuthRefreshStore,
    idempotency: AuthIdempotencyStore,
    jwt_secret: str,
    now: Callable[[], datetime] | None = None,
) -> AuthService:
    return AuthService(
        users=users,
        refresh_store=refresh,
        idempotency_store=idempotency,
        jwt_secret=jwt_secret,
        now=now or (lambda: datetime.now(UTC).replace(tzinfo=None)),
    )
